### PR TITLE
[operator] Using Container Lifecycle Hooks and Inflight Changes for Better Ordering and Delaying Restarts

### LINKED ISSUE #397: bug: Slow Reconciliation of Paladin Nodes and Zeto On-Chain Orchestration
### What happened?

Currently its taking 20-30min for the operator to standup the 3 Besu node network, the accompanying Paladin nodes, and then deploy and invoke the various smart contracts necessary for Noto, Zeto, and Pente.

### What did you expect to happen?

We would like this to ideally take only 1-2m w/ 5m being an acceptable first improvement if possible.

We want users to have a fast and smooth getting started experience, despite the complexities of the multi-node network and demonstrating programmable privacy.

We also want developers to have an E2E pipeline which allows for faster feedback and iteration times.

### How can we reproduce it (as minimally and precisely as possible)?

Follow "Getting Started" or use `gradle deploy` locally.

### Anything else we need to know?

It is believe these are primarily optimizations that need to be made within the Operator, with the exception of the ability to have dynamic reload of registry and domain configuration by Paladin itself. The latter will be treated as a separate issue / body of work, and the Operator will hack around the lack of dynamic reload for now.

### OS version

```console
sw_vers
ProductName:		macOS
ProductVersion:		14.4.1
BuildVersion:		23E224
```

### PR DESCRIPTION
**Pull request by onelapahead**
## Description

Follow up from #402. Attempts to close #397.

Initially on Getting Started, Paladin crashes some due to the Postgres sidecar not being healthy by the time the automigrations run. We order the containers so Postgres is the first container, and use a `postStart` hook that must complete before starting Paladin.

Probes are added so the Operator isn't notified Paladin is ready until its RPC port is live. This should help the deploy/invoke controllers so that they don't run unless it can actually receive JSONRPC traffic.

Then, to avoid the constantly restarting due to changing config / deploying smart contracts issue, we introduce a temporary solution of `inflight` changes map lock. This allows us to detect config changes based on the checksum, delay updating the STS until a change interval of 30s has passed, and then release the changes lock allowing the STS to update, which helps initial SCDs -> TIs -> Domains/Registries all succeed. This can be removed once #398 is completed.

## Other Changes

* Fixes the default container annotation as it must be on the pod
* Include VSCode launch config for running the Operator locally in the debugger
* Increase parallelization of reconcilers (note the internal workqueue of client-go / controller-runtime ensures parallel reconciles of the _same_ resource is not possible).
* Expontential backoff of `setupChainId` in the RPC client of Paladin to avoid any startup crashes due to Besu coming up in parallel (or being down for production scenarios)

### CODE REVIEW

**Comment by onelapahead** [comment]
> Then, to avoid the constantly restarting due to changing config / deploying smart contracts issue, we introduce a bad hack via preStop that tells Paladin to sleep for 15s before shutting down (but the readiness probe continues to pass, allowing traffic to keep being received) which will allow it to deploy more contracts, invoke transactions, and "queue up" other config updates to hopefully help with the delays we've seen.

This isn''t working as desired, as the STS goes into a non-ready state, causing the SCDs/TIs to wait again. Instead, I'll build an inflightChange map to track when a hash has changed, but delay actually reconciling for a few seconds.

File: .vscode/launch.json
Comment (onelapahead) [inline .vscode/launch.json:17]: Note - this assumes you're pointing your operator at `default` per a `gradle deploy`-esque local workflow.

File: .vscode/settings.json
Comment (onelapahead) [inline .vscode/settings.json:4]: On my (newly installed) VSCode w/ Copilot, I was having horrendous performance with basic code inspection and autocomplete w/o upgrading my `golangci-lint` and including these settings.

File: operator/internal/controller/paladin_controller.go
Comment (onelapahead) [inline operator/internal/controller/paladin_controller.go:534]: TODO: need to make this index conditional / contexted so that we can determine if its `1` or `0` depending of PG sidecare mode is enabled.

**Comment by onelapahead** [comment]
> This allows us to detect config changes based on the checksum, delay updating the STS until a change interval of 30s has passed, and then release the changes lock allowing the STS to update, which helps initial SCDs -> TIs -> Domains/Registries all succeed.

Note - this is far from perfect. This does not "lock" the actual change detected, it just guarantees it won't apply any changes but every 30s. Worse case - a final change or two are queued up right after a change is applied via a pod restart, and those then cause a subsequent 30s restart.

But again an overall start up time of 1-3m compared to the previous 5-20m is a vast improvement. This become significantly less complex and faster once dynamic reload is implemented, and therefore feels like an acceptable workaround for the time being, but will defer to @dwertent @peterbroadhurst .

File: operator/internal/controller/check-psql.sh
Comment (dwertent) [inline operator/internal/controller/check-psql.sh:8]:
```suggestion
  sleep 1
```

Comment (dwertent) [reply operator/internal/controller/check-psql.sh:8]: I think this will save some startup time

File: operator/Makefile
Comment (dwertent) [inline operator/Makefile:180]: This ensures we do not copy files that are not CRDs. Is there a reason to remove it?

File: operator/hack/check-running-on-kind.sh
Comment (dwertent) [inline operator/hack/check-running-on-kind.sh:1]: What is this for?

File: operator/internal/controller/inflight.go
Comment (dwertent) [inline operator/internal/controller/inflight.go:11]: I highly recommend using the [maps](https://github.com/goradd/maps) package (`maps.SafeMap`), for asynchronous map operations

**Comment by jimthematrix** [comment]
Testing with the branch, I saw much reduced crashloopfallback status in the paladin pods (where before the fix I saw all 3 paladin pods going through the crashloop before things get resolved/reconciled). even when I do, things get resolved within a couple of seconds, and the entire bootstrapping process was complete in under 60s for me on an M1 MacBook Pro.

File: operator/internal/controller/check-psql.sh
Comment (onelapahead) [reply operator/internal/controller/check-psql.sh:8]: Yeah its tricky... the Postgres container temporarily starts its Postgres server to do some bootstrapping, shuts down, and then starts back up again.

So I've found the time needs to be longer in order to not accidentally consider Postgres "ready" too soon.

I wonder if instead I can look for the presence of a folder/file to indicate bootstrapping is done, before doing a quick health check.

File: operator/hack/check-running-on-kind.sh
Comment (onelapahead) [reply operator/hack/check-running-on-kind.sh:1]: If you see the VSCode launch config for running the operator in an IDE and the check task - its used by my launch config to prevent a dev from accidentally running the Paladin operator against an unintended cluster. As an SRE working across a lot of clusters daily, I appreciate when projects have safeguards in place for local development.

File: operator/Makefile
Comment (onelapahead) [reply operator/Makefile:180]: I'm using Kustomize itself to target the `config/crd` kustomization folder (not `config/default`) which therefore only output the CRDs, avoiding the need to delete extra YAML files.

File: operator/internal/controller/check-psql.sh
Comment (dwertent) [reply operator/internal/controller/check-psql.sh:8]: Interesting. Maybe it's worth querying `pg_isready`? I believe that should respond only once the server is ready

**Review by dwertent — APPROVED** [review]
Well done!

File: operator/internal/controller/paladin_controller.go
Comment (onelapahead) [reply operator/internal/controller/paladin_controller.go:542]: Discussed with @dwertent - this code should just "find" the container rather than guess - in case we add any future sidecars (hopefully spinning the UI out separately).

File: operator/internal/controller/check-psql.sh
Comment (onelapahead) [reply operator/internal/controller/check-psql.sh]: Great suggestion - testing that now

File: operator/internal/controller/inflight.go
Comment (onelapahead) [reply operator/internal/controller/inflight.go:11]: Going to hold off on this for now just to avoid extra deps and bc this concurrent map is simple, but will keep that in mind for the future 🙇🏻

**Pull request merged**
